enhancements/sig-architecture/70-spread-policy/README.md

```diff
@@ -0,0 +1,274 @@
+}
+```
+
+### API Example
+
+#### Example 1: Joint use of a Even constraint and Affinity constraints
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: Placement
+metadata:
+  name: demo-placement
+spec:
+  # numberOfClusters is required if spreadConstraints contains
+  # a constraint of Even type.
+  numberOfClusters: 10
+  # Prediates work before the spread constraints to filter
+  # out some undesired topologies.
+  predicates:
+  - # ...
+  # SpreadConstraints contain two constraints, which work
+  # independently and jointly.
+  spreadConstraints:
+    # order is omitted for this Even constraint as there is only
+    # one of it.
+  - type: Even
+    topologyKey: provider
+    topologyKeyType: Label
+    maxSkew: 2
+    # Affininy constraints do not have a maxSkew field.
+  - type: Affinity
+    topologyKey: region
+    topologyKeyType: Label
+    # The cluster affinity api doesn't use LabelSelector
+    # directly in order to force users to organize the
+    # affinity rules of the same topologyKey together,
+    # which may be more clear.
+    topologyWeights:
+    - weight: 20
+      operator: In
+      values:
+      - us-east-1
+      - cn-hangzhou
+    - weight: 10
+      operator: In
+      values:
+      - us-west-1
+    - weight: 20
+      # NotIn expresses anti-affinity.
+      operator: NotIn
+      values:
+      - cn-shenzhen
+  - type: Affinity
+    topologyKey: zone
+    topologyKeyType: Label
+    topologyWeights:
+    - # ...
+```
+
+#### Example 2: Joint use of multiple Even rules
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: Placement
+metadata:
+  name: demo-placement
+spec:
+  numberOfClusters: 10
+  predicates:
+  - # ...
+  spreadConstraints:
+    # When multiple Even constraints are used, the user should
+    # specify their orders. The scheduler will consider to first
+    # do the spread according to the Even constraints with smaller
+    # order value.
+  - type: Even
+    topologyKey: provider
+    topologyKeyType: Label
+    maxSkew: 2
+    order: 1
+    # maxSkew is omitted as this one is a soft Even constraint
+  - type: Even
+    topologyKey: region
+    topologyKeyType: Label
+    order: 2
+```
+
+### Implementation
+Two types of spread constraints require different logics. Hence, they will be implemented separately. Affinity constraints are implemented by a `Prioritizer` to calculate the scores of the clusters using the information conveyed via `topologyWeights`. However, it seems that the scheduler's current design of the plugin framework makes it difficult to express the logic of even spreads. Therefore, an extra `Selector` plugin type is proposed.
+
+**Implementation of (Cluster) Affinity/Anti-affinity.** An `AffinitySpreadPriorititizer` will be implemented as follows:
+
+1. for each cluster, score it according to `topologyWeights` in the cluster affinity rules
+2. normalize the cluster affinity scores to [0, 100]
```

Comment:
How will this score work together with other Prioritizers? For example:
```
spreadConstraints:
  - type: Affinity
    topologyKey: region
    topologyKeyType: Label
    topologyWeights:
    - weight: 20
      operator: In
      values:
      - us-east-1
      - cn-hangzhou
    - weight: 10
      operator: In
      values:
      - us-west-1
 prioritizerPolicy:
    mode: Exact
    configurations:
      - scoreCoordinate:
          builtIn: ResourceAllocatableMemory
```
Will cluster in us-west-1 with much more ResourceAllocatableMemory has a higher score than cluster in us-east-1?

Reply:
I consider not to include the Affinity and Even constraints in Additive mode. In Exact mode, users may need to specify the weight of them manually. The above example that only contains affinity constraints will be like:

```yaml
spreadConstraints:
  - type: Affinity
    topologyKey: region
    topologyKeyType: Label
    topologyWeights:
    - weight: 20
      operator: In
      values:
      - us-east-1
      - cn-hangzhou
    - weight: 10
      operator: In
      values:
      - us-west-1
 prioritizerPolicy:
    mode: Exact
    configurations:
      - scoreCoordinate:
          builtIn: ResourceAllocatableMemory
      - scoreCoordinate:
          builtIn: ClusterAffinitySpread
        weight: 2
```

In this scenario, if we only have one cluster in each region mentioned above (i.e., cluster A in us-west-1 and cluster B in us-est-1), then the affinity prioritizer will score cluster A to be 100 and cluster B to be 50 after normalization. If the memory resource score of cluster A and cluster B are α and β, the final score of them will be α+200 (for cluster A) and β+100 (for cluster B).

The proposal will be updated soon to contain examples about the weight of Affinity and Even.